### ISSUE TITLE
kSecUseAuthenticationContext Support

### ISSUE THREAD

**Issue by juantrias** [square/Valet#104]
We are developing an application that stores an encryption key on a `VALSecureEnclaveValet` protected by `VALAccessControlTouchIDCurrentFingerprintSet`. We want to protect the key with TouchID but we do not want to request the TouchID everytime we need the key because we use it at almost every screen of our app to decrypt the data displayed on screen. We want to request TouchID once and keep access to the key "authenticated" while our session is alive. Keeping the key in memory for a long time is not an option for us.

We have achieved the behavior we want passing a LAContext object at `kSecUseAuthenticationContext` at the `[VALValet stringForKey:options:] ` options dictionary. But this is a protected method and we cannot call it without modifying your library code. Do you plan to add the possibility to pass a LAContext to `[VALSecureEnclaveValet stringForKey]`? Or could you suggest me a better way to implement this functionality?

Thanks

**Comment by dfed** [comment]
As a short-term hack, you may be able to write an extension that declares `stringForKey:options:` on `VALValet` in your code. We do not have any desire to expose the `options` dictionary call, since that's an invitation to abuse of our API, and is more confusing than helpful to the average Valet adoptee.

Longer term, this sounds like a feature worth adding to Valet. That said, I don't have a need for it personally, so you're going to need to give me some help. To give me better context on how this API is used, can you attach a diff or a code sample of how you use `stringForKey:options:` and how you create/evaluate an `LAContext` object?

Previously, we've been loathe to expose any underlying Security API, since it's quite hard to grok. If `LAContext` is easy to reason about, we may add `stringForKey:userPrompt:authenticationContext:userCancelled:` and `objectForKey:userPrompt:authenticationContext:userCancelled:` methods to `VALSecureEnclaveValet` and handle the options dictionary under the hood. If `LAContext` is difficult to use... we may want to add a `VALContextualSecureEnclaveValet` (name suggestions welcome) class that encapsulates this behavior.

**Comment by juantrias** [comment]
I have wrote a `VALSecureEnclaveValet` extension that declares a `stringForKey:userPrompt:authenticationContext:userCancelled` method. I've had to include the Pod private headers in my project to be able to `#import "VALValet_Protected.h"` and call `[VALValet stringForKey:options:status]`

About the convenience of having a dedicated `VALContextualSecureEnclaveValet` subclass I don't have a definitive opinion. I don't think that `LAContext` is difficult to use, but it has its peculiarities. For instance when calling `SecItemCopy` passing the same LAContext multiple times it continues returning the "cached" value even if the item has been deleted from the Keychain due to fingerprint changes. If don't know if this is the expected behavior or a bug, but I have reproduced it on the [WWDC Keychain sample app](https://developer.apple.com/library/content/samplecode/KeychainTouchID/Introduction/Intro.html), modifying the code to pass an `LAContext`. 

This is why in my project I have decided to encapsulate `VALSecureEnclaveValet` in a wrapper that manages the `LAContext` internally.

I share with you both the extension and the wrapper: https://gist.github.com/juantrias/193c94e0fc10ded628fd5c8df5ad3b71
